### PR TITLE
Coverage report generation for evmzero

### PR DESCRIPTION
**Pull request by LuisPH3**
This PR is part of #519, 

This iteration includes a coverage report tool which can report coverage of the code executed by the C++ unit tests.
Unfortunately, this approach works best when using GCC.

The use is described in the build.md file.
- compile instrumented code
- run tests
- process profile data and generate report.

### CODE REVIEW

File: BUILD.md [hunk 1]
```diff
@@ -216,4 +216,35 @@ Fib/20/lfvm-no-code-cache-12  21.2ms ± 3%  20.7ms ± 3%  -2.17%  (p=0.006 n=9+1
...
\ No newline at end of file
+Ideally, pull requests targeting performance improvements should include such a report in their description to document its impact.
+
+## Code Coverage
+
+Code coverage of the Tosca project is a work in progress. As the different systems get integrated, more features will be added and documented here.
+
+### C++ Code coverage
+
+Current code coverage infrastructure uses GNU gcov system. This is supported in both Gcc and Clang, although Clang target gcov version may not be the installed one. For this reason the current infrastructure is currently enabled for GCC only.
```
Comment (facuMH) [inline BUILD.md:227]:
```suggestion
Current code coverage infrastructure uses GNU gcov system. This is supported in both GCC and Clang, although Clang target gcov version may not be the installed one. For this reason the current infrastructure is currently enabled for GCC only. 
```

File: BUILD.md [hunk 2]
```diff
@@ -216,4 +216,35 @@ Fib/20/lfvm-no-code-cache-12  21.2ms ± 3%  20.7ms ± 3%  -2.17%  (p=0.006 n=9+1
...
+Current code coverage infrastructure uses GNU gcov system. This is supported in both Gcc and Clang, although Clang target gcov version may not be the installed one. For this reason the current infrastructure is currently enabled for GCC only. 
+
+Code coverage depends on the following packages:
+- lcov
+- genhtml
+Both commands can be installed in ubuntu with the command: `apt install lcov`
+
+Code coverage can be triggered by using the following command:
+```
+CC=gcc CXX=g++ make test-cpp-coverage
```
Comment (facuMH) [inline BUILD.md:236]: Thanks for adding it directly in the main make !!

File: cpp/CMakeLists.txt [hunk 3]
```diff
@@ -2,8 +2,9 @@ cmake_minimum_required(VERSION 3.19)
 set(CMAKE_POLICY_DEFAULT_CMP0077 NEW) # option respects set variables
 project(tosca)
 
-option(TOSCA_ASAN "Enable AddressSanitizer for Tosca targets.")
-option(TOSCA_ASSERT "Enable Tosca specific assertions.")
+option(TOSCA_ASAN "Enable AddressSanitizer for Tosca targets." OFF)
```
Comment (facuMH) [inline cpp/CMakeLists.txt:5]: why is default off ?

File: cpp/CMakeLists.txt [hunk 4]
```diff
@@ -49,7 +84,7 @@ function(tosca_add_compile_flags target)
     CXX_EXTENSIONS OFF)
   target_compile_options(${target} PRIVATE
     $<$<CXX_COMPILER_ID:MSVC>:/MP /W4 /external:W0>
-    $<$<CXX_COMPILER_ID:GNU,Clang,AppleClang>:-march=native -mtune=native -Wall -Wextra -Wpedantic -Wtype-limits -Wconversion -Wsign-conversion -Wdouble-promotion -g>
```
Comment (facuMH) [inline cpp/CMakeLists.txt:52]: why remove `-g` flag ?

**Review by facuMH — DISMISSED** [review]
Looks great, thanks !! just a couple of questions out of curiosity

File: cpp/CMakeLists.txt (hunk 3 above)
Comment (LuisPH3) [reply cpp/CMakeLists.txt:5]: I don't want to instrument code by default. 
The default build shall be not-instrumented  (either sanitizer nor coverage) 
(this defaults only take effect if you call CMake directly, the top level Makefile will always pre-fill them.

File: cpp/CMakeLists.txt (hunk 4 above)
Comment (LuisPH3) [reply cpp/CMakeLists.txt:52]: This kind of controversial, I will revert it.
The thing is that debug symbols were always enabled, even for release builds. This flag would be duplicated in some cases... 
I do not need debug symbols for release builds, and IMO if needed, should be optional. I will come back here to clean this flags some other day.

**Review by facuMH — DISMISSED** [review]

File: BUILD.md [hunk 5]
```diff
@@ -216,4 +217,35 @@ Fib/20/lfvm-no-code-cache-12  21.2ms ± 3%  20.7ms ± 3%  -2.17%  (p=0.006 n=9+1
...
+Current code coverage infrastructure uses GNU gcov system. This is supported in both Gcc and Clang, although Clang target gcov version may not be the installed one. For this reason the current infrastructure is currently enabled for Gcc only. 
+
+Code coverage depends on the following packages:
+- lcov
+- genhtml
+Both commands can be installed in Ubuntu with the command: `apt install lcov`
+
+Code coverage can be triggered by using the following command:
+```
+CC=gcc CXX=g++ make test-cpp-coverage
```
Comment (HerbertJordan) [inline BUILD.md:237]: I tried to run this command, but I got this error:
```
$ CC=gcc CXX=g++ make test-cpp-coverage
cd cpp ; \
cmake \
	-Bbuild \
	-DCMAKE_BUILD_TYPE="Debug" \
	-DCMAKE_SHARED_LIBRARY_SUFFIX_CXX=.so \
	-DTOSCA_ASSERT="ON" \
	-DTOSCA_COVERAGE="ON" \
	-DTOSCA_ASAN="OFF"; \
cmake --build build --parallel
CMake Error at CMakeLists.txt:28 (message):
  Coverage is only supported with GCC.
```

After a while I realized that the `cmake` build was already configured, and it is ignoring the new build. Running
```
make clean
```
fixed the problem.

Should this be mentioned here somewhere?

File: Makefile [hunk 6]
```diff
@@ -44,6 +50,13 @@ test-cpp-asan: TOSCA_CPP_BUILD = Debug
 test-cpp-asan: TOSCA_CPP_ASAN = ON
 test-cpp-asan: test-cpp
 
+test-cpp-coverage: TOSCA_CPP_BUILD = Debug
+test-cpp-coverage: TOSCA_CPP_COVERAGE = ON
+test-cpp-coverage: test-cpp
```
Comment (HerbertJordan) [inline Makefile:55]: I am terrible with Makefiles, but isn't this dependency on `test-cpp` cause the tests to be run twice?

File: cpp/CMakeLists.txt [hunk 7]
```diff
@@ -2,8 +2,9 @@ cmake_minimum_required(VERSION 3.19)
 set(CMAKE_POLICY_DEFAULT_CMP0077 NEW) # option respects set variables
 project(tosca)
 
-option(TOSCA_ASAN "Enable AddressSanitizer for Tosca targets.")
-option(TOSCA_ASSERT "Enable Tosca specific assertions.")
+option(TOSCA_ASAN "Enable AddressSanitizer for Tosca targets." OFF)
+option(TOSCA_ASSERT "Enable Tosca specific assertions." ON)
```
Comment (HerbertJordan) [inline cpp/CMakeLists.txt:6]: The assertions should be off in the release build (target `tosca-cpp`). Is it like that?

File: cpp/CMakeLists.txt [hunk 8]
```diff
@@ -12,10 +13,43 @@ enable_testing()
...
+
+  find_program(LCOV lcov REQUIRED)
+  find_program(GENHTML genhtml REQUIRED)
+
+  add_custom_target(coverage
+    COMMENT "Generating coverage report."
+    COMMAND ${LCOV} 
+      --capture 
+      --directory .  
+      --ignore-errors mismatch
```
Comment (HerbertJordan) [inline cpp/CMakeLists.txt:42]: I got an error from this flag:
```
Generating coverage report.
Capturing coverage data from .
Subroutine read_intermediate_text redefined at /usr/bin/geninfo line 2623.
Subroutine read_intermediate_json redefined at /usr/bin/geninfo line 2655.
Subroutine intermediate_text_to_info redefined at /usr/bin/geninfo line 2703.
Subroutine intermediate_json_to_info redefined at /usr/bin/geninfo line 2792.
Subroutine get_output_fd redefined at /usr/bin/geninfo line 2872.
Subroutine print_gcov_warnings redefined at /usr/bin/geninfo line 2900.
Subroutine process_intermediate redefined at /usr/bin/geninfo line 2930.
Found gcov version: 11.4.0
Using intermediate gcov format
geninfo: ERROR: unknown argument for --ignore-errors: mismatch
gmake[4]: *** [CMakeFiles/coverage.dir/build.make:71: CMakeFiles/coverage] Error 255
gmake[4]: Leaving directory '/home/herbert/coding/fantom/tosca/cpp/build'
gmake[3]: *** [CMakeFiles/Makefile2:1048: CMakeFiles/coverage.dir/all] Error 2
gmake[3]: Leaving directory '/home/herbert/coding/fantom/tosca/cpp/build'
gmake[2]: *** [CMakeFiles/Makefile2:1055: CMakeFiles/coverage.dir/rule] Error 2
gmake[2]: Leaving directory '/home/herbert/coding/fantom/tosca/cpp/build'
gmake[1]: *** [Makefile:200: coverage] Error 2
gmake[1]: Leaving directory '/home/herbert/coding/fantom/tosca/cpp/build'
make: *** [Makefile:56: test-cpp-coverage] Error 2
```

I have LCOV version 1.14.

**Review by HerbertJordan — DISMISSED** [review]

File: BUILD.md (hunk 5 above)
Comment (LuisPH3) [reply BUILD.md:237]: Sorry, this is a no-brainer to me. CMake cannot change toolchain once configured. I'll write it down.

File: cpp/CMakeLists.txt (hunk 7 above)
Comment (LuisPH3) [reply cpp/CMakeLists.txt:6]: Same defaults as in makefile, makefile values take precedence. 
image
if this is wrong, let me know.

File: cpp/CMakeLists.txt (hunk 8 above)
Comment (LuisPH3) [reply cpp/CMakeLists.txt:42]: ubuntu 20? 22? 
can you please run  `dpkg -S $(which lcov)`

File: Makefile (hunk 6 above)
Comment (LuisPH3) [reply Makefile:55]: I haven't written a makefile in 20 years... yes, the unit test was running twice.

File: cpp/CMakeLists.txt (hunk 8 above)
Comment (simonlechner) [reply cpp/CMakeLists.txt:42]: As expected, I am using the same Ubuntu version (22.04.) as @HerbertJordan and I get the same error.
Running the `dpkg` command return `/usr/bin/lcov`.

Comment (LuisPH3) [reply cpp/CMakeLists.txt:42]: Ok, problem comes from Gcc 11 vs 13:

1. gcc 11 does not have this ignore feature. This is default in ubuntu 22
2. gcc 13 complains about it and requires the argument. This is default in ubuntu 24

will see what can I do about it.

**Review by facuMH — APPROVED** [review]
runs smoothly with ubu 24

File: cpp/CMakeLists.txt (hunk 7 above)
Comment (HerbertJordan) [reply cpp/CMakeLists.txt:6]: We should discuss the role of assertions in the C++ code offline some time and then update those flags if needed. So, its beyond the scope of this PR.

File: cpp/CMakeLists.txt (hunk 8 above)
Comment (LuisPH3) [reply cpp/CMakeLists.txt]: should work now.

**Review by HerbertJordan — DISMISSED** [review]
Thanks for the fixes, works for me now!

**Review by facuMH — DISMISSED** [review]

File: BUILD.md [hunk 9]
```diff
@@ -198,7 +199,7 @@ This package makes C/C++ symbols accessible to the Go profiler. On some systems,
 
 To compare the benchmark results of two different code versions, the [benchstat](https://pkg.go.dev/golang.org/x/perf/cmd/benchstat) tool can be utilized. To install the tool run
 
-```
+```bash
```
Comment (simonlechner) [inline BUILD.md:202]: Is there a reason why you used `sh` for all cases except this one?

**Review by simonlechner — DISMISSED** [review]
Thank you for reworking it, also works on my machine now.

File: BUILD.md (hunk 9 above)
Comment (LuisPH3) [reply BUILD.md:202]: not really.

**Review by simonlechner — APPROVED** [review]

**Pull request merged**